src/java.base/share/classes/jdk/internal/vm/ThreadContainer.java

```diff
@@ -76,24 +76,47 @@ public long threadCount() {
     public abstract Stream<Thread> threads();
 
     /**
-     * Invoked by Thread::start before the given Thread is started.
+     * Invoked by {@code add} to add a thread to this container before it starts.
      */
-    public void onStart(Thread thread) {
-        // do nothing
+    protected void onStart(Thread thread) {
     }
 
     /**
-     * Invoked when a Thread terminates or starting it fails.
-     *
-     * For a platform thread, this method is invoked by the thread itself when it
-     * terminates. For a virtual thread, this method is invoked on its carrier
-     * after the virtual thread has terminated.
-     *
-     * If starting the Thread failed then this method is invoked on the thread
-     * that invoked onStart.
+     * Invoked by {@code remove} to remove a thread from this container when it
+     * terminates (or failed to start).
      */
-    public void onExit(Thread thread) {
-        // do nothing
+    protected void onExit(Thread thread) {
+    }
+
+    /**
+     * Adds a thread to this container. This method should be invoked before the
+     * thread executes.
+     */
+    public final void add(Thread thread) {
+        // Prevent a virtual thread from being preempted as this could potentially
+        // deadlock with a carrier is removing a virtual thread from the container
```

Comment:
Is there a typo in this sentence? Should "with a carrier" have been "when a carrier ..." ? Similar comment in the remove(...) method too.

Reply:
You are right, there is a pre-existing typo in that comment (the sentence was copied to here from RQ).